### PR TITLE
[RELAY] BiasAdd, MLP, Resnet testing

### PR DESCRIPTION
**Pull request by tqchen**
This PR recovers the testing API to construct the normal workloads in relay. It contains the following changes

- [x] Refactor ir_pass.free_vars to return vars in post-dfs order @MarisaKirisame 
   - This allows us to directly use this to list unbound vars, which corresponds to old nnvm.symbol.list_arguments
- [x] Enable batch_norm to infer the type of the gamma, beta, moving_mean @slyubomirsky 
- [x] Disable node generic inheritance TupleWrapper
   - This is because batch_norm API now returns a tuple, which differs from traditional batch_norm in the inference API
   - This is a reasonable change as it makes batch_norm functional and opens for more optimizations
   - The decision is to allow the system to eagerly throw when TupleWrapper is mistakenly passed into operators
   - For cases when tuple is needed, astuple must be explicitly called.
- [x] Introduce relay.testing, which now includes a mlp and resnet

### CODE REVIEW

**Comment by tqchen** [comment]
cc @joshpoll @jroesch @yzhliu @zhiics  please review

File: python/tvm/relay/expr.py [hunk 1]
```diff
@@ -275,6 +296,9 @@ def __repr__(self):
         return ("TupleWrapper(" + self.tuple_value.__repr__() +
                 ", " + self.size + ")")
 
+    def astype(self, _):
+        raise TypeError("astype cannot be used on tuple")
```
Comment (MarisaKirisame) [inline python/tvm/relay/expr.py]: tuple -> tuplewrapper?

File: python/tvm/relay/expr.py [hunk 2]
```diff
@@ -28,6 +28,25 @@ def checked_type(self):
...
+        dtype : str
+            The target data type.
+
+        Note
+        ----
+        This function only works for TensorType Exprs.
+
+        Returns
+        -------
+        result : relay.Expr
```
Comment (MarisaKirisame) [inline python/tvm/relay/expr.py]: tvm.relay.Expr

File: python/tvm/relay/op/nn/nn.py [hunk 3]
```diff
@@ -425,12 +426,38 @@ def batch_flatten(data):
...
+
+    Parameters
+    ----------
+    data : relay.Expr
+        The input data to the operator.
+
+    bias : relay.Expr
+        The bias to be added.
+
+    axis : int, optional
```
Comment (MarisaKirisame) [inline python/tvm/relay/op/nn/nn.py]: Optional[int]

File: python/tvm/relay/testing/init.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+    Parameters
+    ----------
+    rnd_type: str, optional
+        Random generator type, can be ``'gaussian'`` or ``'uniform'``.
+
+    factor_type: str, optional
+        Can be ``'avg'``, ``'in'``, or ``'out'``.
+
+    magnitude: float, optional
```
Comment (MarisaKirisame) [inline python/tvm/relay/testing/init.py]: Optional[float]

File: python/tvm/relay/testing/init.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+class Xavier(Initializer):
+    """ "Xavier" initialization for weights
+
+    Parameters
+    ----------
+    rnd_type: str, optional
+        Random generator type, can be ``'gaussian'`` or ``'uniform'``.
+
+    factor_type: str, optional
```
Comment (MarisaKirisame) [inline python/tvm/relay/testing/init.py]: Optional[str]

File: python/tvm/relay/testing/init.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        else:
+            raise ValueError("Unknown random type")
+
+
+def create_workload(net, initializer=None, seed=0):
+    """Helper function to create benchmark image classification workload.
+
+    Parameters
+    ----------
+    net : relay.Function
```
Comment (MarisaKirisame) [inline python/tvm/relay/testing/init.py]: tvm.relay.Function

File: python/tvm/relay/testing/layers.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+                               moving_var=moving_var,
+                               **kwargs)[0]
+
+
+def conv2d(data, weight=None, **kwargs):
+    """Wrapper of conv2d which automatically creates weights if not given.
+
+    Parameters
+    ----------
+    data : relay.Expr
```
Comment (MarisaKirisame) [inline python/tvm/relay/testing/layers.py]: tvm.relay.Expr

File: python/tvm/relay/testing/layers.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+def dense_add_bias(data, weight=None, bias=None, **kwargs):
+    """Wrapper of dense which automatically creates weights if not given.
+
+    Parameters
+    ----------
+    data : relay.Expr
+        The input expression.
+
+    weight : relay.Expr
```
Comment (MarisaKirisame) [inline python/tvm/relay/testing/layers.py]: tvm.relay.Expr

File: python/tvm/relay/testing/mlp.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    batch_size : int
+        The batch size used in the model
+
+    num_classes : int, optional
+        Number of claseses
+
+    image_shape : tuple, optional
+        The input image shape
+
+    dtype : str, optional
```
Comment (MarisaKirisame) [inline python/tvm/relay/testing/mlp.py]: Optional[str]

File: python/tvm/relay/testing/resnet.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+    ----------
+    data : str
+        Input data
+    num_filter : int
+        Number of output channels
+    bnf : int
+        Bottle neck channels factor with regard to num_filter
+    stride : tuple
+        Stride used in convolution
+    dim_match : Boolean
```
Comment (MarisaKirisame) [inline python/tvm/relay/testing/resnet.py]: bool

File: include/tvm/relay/pass.h [hunk 11]
```diff
@@ -102,35 +102,26 @@ bool AlphaEqual(const Type& t1, const Type& t2);
  */
 bool WellFormed(const Expr& e);
 
-/*! \brief Get free variables from expression e.
+/*! \brief Get free Vars from expression expr in PostDFS order.
```
Comment (jroesch) [inline include/tvm/relay/pass.h]: Is `expression expr` a typo?

**Comment by jroesch** [comment]
Looks good to me 👍, removed similar code from the RTS PR.

**Comment by tqchen** [comment]
Thanks @jroesch @MarisaKirisame for reviews, this is now merged

File: python/tvm/relay/testing/init.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+            factor = (fan_in + fan_out) / 2.0
+        elif self.factor_type == "in":
+            factor = fan_in
+        elif self.factor_type == "out":
+            factor = fan_out
+        else:
+            raise ValueError("Incorrect factor type")
+        # Hack for mobilenet, because there is less connectivity
+        if "depthwise" in name:
+            factor = 3 * 3
```
Comment (merrymercy) [inline python/tvm/relay/testing/init.py]: Should we delete this?

File: python/tvm/relay/testing/resnet.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+                                      name=name + '_bn1')
+        act1 = relay.relu(data=bn1)
+        conv1 = layers.conv2d(
+            data=act1,
+            channels=int(num_filter*0.25),
+            kernel_size=(1, 1),
+            strides=stride,
+            padding=(0, 0),
+            name=name + '_conv1')
+        bn2 = layers.batch_norm_infer(data=conv1, epsilon=2e-5, name=name + '_bn2')
```
Comment (merrymercy) [inline python/tvm/relay/testing/resnet.py]: These APIs are not as consistent as the old ones in nnvm. Some start with `layers.`, some start with `relu.`, and others start with `relu.nn`. Some ends with `_infer`, while the others do not.

We should provide a more consistent API for front-end users. 
In this front-end api, we can hide all variable declerations and things like `bias_add`

Comment (tqchen) [reply python/tvm/relay/testing/resnet.py]: True, so far the layers API is a simple extension to the original API and things are for automatic variable creation. As an IR and compiler, we don't need a deep layered API, but I agree that we can use a better API(like gluon) if we want to present a frontend.

Comment (eqy) [reply python/tvm/relay/testing/resnet.py]: I am looking at porting mobilenet; one main difference is that it seems before we named relu ops in the graph, but now `relay.relu` does not take in a name parameter. Not sure if this is a big deal (it seems conv2d and batch norm have wrappers that take in a name parameter).

File: src/relay/op/nn/nn.cc [hunk 14]
```diff
@@ -371,58 +437,26 @@ The whole array is rescaled by ``1/(1-p)`` to keep the expected sum of the input
...
-}
-
 bool BatchNormRel(const Array<Type>& types,
                   int num_inputs,
                   const Attrs& attrs,
                   const TypeReporter& reporter) {
   CHECK_EQ(types.size(), 6);
   const auto* data = types[0].as<TensorTypeNode>();
   if (data == nullptr) return false;
-  if (data->shape.size() == 0) return false;
```
Comment (slyubomirsky) [inline src/relay/op/nn/nn.cc]: Doesn't this mean that we get an out of bounds if the axis is -1 and the data shape is of rank 0? In that case, `axis` would be equal to `data->shape.size() - 1`, or -1, so on line 453, `axis_size` would be `data->shape[-1]`. (Sorry for not reviewing before it was merged.)

Comment (tqchen) [reply src/relay/op/nn/nn.cc]: This is a good catch, I think it is better to directly do a boundary checking (after axis is transformed back by axis + data->shape.size() to avoid this problem

File: python/tvm/relay/testing/resnet.py (hunk 13 above)
Comment (tqchen) [reply python/tvm/relay/testing/resnet.py]: The name parameter is used to construct weight's name. For now, it is not a big problem in our end
